### PR TITLE
VSTS CLI Release cmds added

### PR DESCRIPTION
**Pull request by nishubansal**

### CODE REVIEW

**Comment by vbossica** [comment]
Would it be possible to add methods to download and upload a release definition?

File: src/command_modules/vsts-cli-release/vsts/cli/vrelease/commands.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+from ._format import (transform_release_table_output,
+                      transform_releases_table_output,
+                      transform_release_definition_table_output,
+                      transform_release_definitions_table_output)
+from knack.commands import CommandGroup
+
+
+def load_release_commands(cli_command_loader):
+    with CommandGroup(cli_command_loader, 'release', 'vsts.cli.vrelease.common.{}') as g:
```
Comment (atbagga) [inline src/command_modules/vsts-cli-release/vsts/cli/vrelease/commands.py]: Typo - vrelease?
```suggestion
    with CommandGroup(cli_command_loader, 'release', 'vsts.cli.release.common.{}') as g:
```

File: src/common_modules/vsts-cli-release-common/vsts/cli/vrelease/common/release.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+from vsts.cli.common.services import (get_release_client, resolve_instance_and_project)
+from vsts.cli.common.uri import uri_quote
+from .release_definition import get_definition_id_from_name
+
+logger = get_logger(__name__)
+
+def release_create(definition_id=None, definition_name=None, artifact_metadata_list=None, description=None, open_browser=False,
+                team_instance=None, project=None, detect=None):
+    """Request (create) a release.
+    :param definition_id: ID of the definition to create. Required if --name is not supplied.
```
Comment (atbagga) [inline src/common_modules/vsts-cli-release-common/vsts/cli/vrelease/common/release.py]: for short names like --name to work you will need to give the options list in the arguments.py file. 

similar to this.. ac.argument('team_instance', options_list=('--instance', '-i'))

File: src/common_modules/vsts-cli-release-common/vsts/cli/vrelease/common/release_definition.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from webbrowser import open_new
+
+from knack.log import get_logger
+from vsts.cli.common.services import (get_release_client, get_git_client,
```
Comment (atbagga) [inline src/common_modules/vsts-cli-release-common/vsts/cli/vrelease/common/release_definition.py]: get_git_client unused.

File: src/common_modules/vsts-cli-release-common/vsts/cli/vrelease/common/release.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                                status_filter=status)
+    return releases
+
+
+def _open_release(release, team_instance):
+    """Open the release results page in your web browser.
+    :param :class:`<Release> <release.v4_0.models.Release>` release:
+    :param str team_instance:
+    """
+    # https://mseng.visualstudio.com/vsts-cli/_release/index?releaseId=4053990
```
Comment (bansalaseem) [inline src/common_modules/vsts-cli-release-common/vsts/cli/vrelease/common/release.py]: does this url work for release?

File: src/common_modules/vsts-cli-release-common/vsts/cli/vrelease/common/release_definition.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        _open_definition(release_definition, team_instance)
+    return release_definition
+
+
+def _open_definition(definition, team_instance):
+    """Opens the release definition in the default browser.
+    :param :class:`<ReleaseDefinitionReference> <release.v4_0.models.ReleaseDefinitionReference>` definition:
+    :param str team_instance:
+    """
+    # https://mseng.visualstudio.com/vsts-cli/_release/index?definitionId=5419
```
Comment (bansalaseem) [inline src/common_modules/vsts-cli-release-common/vsts/cli/vrelease/common/release_definition.py]: does this url work?

File: scripts/docgen/doc_source_map.json [hunk 6]
```diff
@@ -1,8 +1,9 @@
 {
-    "build": "src/command_modules/vsts-cli-build/vsts/cli/build/_help.py",
```
Comment (gauravsaralMs) [inline scripts/docgen/doc_source_map.json]: avoid changes in lines which are not touched...
seems like whitespace changed in all lines

File: src/command_modules/vsts-cli-release/vsts/cli/vrelease/_format.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    table_row = OrderedDict()
+    table_row['ID'] = row['id']
+    table_row['Name'] = row['name']
+    table_row['Status'] = row['status']
+    table_row['Definition ID'] = row['releaseDefinition']['id']
+    table_row['Definition Name'] = row['releaseDefinition']['name']
+    table_row['Created on'] = row['createdOn']
+    table_row['Reason'] = row['reason']
+    return table_row
+
```
Comment (atbagga) [inline src/command_modules/vsts-cli-release/vsts/cli/vrelease/_format.py]: Table outputs are supposed to be confined such that they do not overflow to a new line on an average resolution screen. 
Bottom line being keep it as short as possible such that it is useful for human user. 

In the above fields I think we can get rid of IDs if we are outputting names. If there are commands which will not accept names but only Ids then you will need to display Ids so user can copy and use them in other commands.

Can you share a sample output for this?

File: src/common_modules/vsts-cli-release-common/vsts/cli/vrelease/common/release.py (hunk 4 above)
Comment (nishubansal) [reply src/common_modules/vsts-cli-release-common/vsts/cli/vrelease/common/release.py]: I have removed the hardcoding and changed the logic to get web url from release object

File: src/common_modules/vsts-cli-release-common/vsts/cli/vrelease/common/release_definition.py (hunk 5 above)
Comment (nishubansal) [reply src/common_modules/vsts-cli-release-common/vsts/cli/vrelease/common/release_definition.py]: same as above.

File: src/command_modules/vsts-cli-release/vsts/cli/vrelease/_format.py (hunk 7 above)
Comment (nishubansal) [reply src/command_modules/vsts-cli-release/vsts/cli/vrelease/_format.py]: sample-output

File: scripts/docgen/doc_source_map.json [hunk 8]
```diff
@@ -1,8 +1,9 @@
...
-    "code": "src/command_modules/vsts-cli-code/vsts/cli/code/_help.py",    
-    "team": "src/command_modules/vsts-cli-team/vsts/cli/team/_help.py",
-    "work": "src/command_modules/vsts-cli-work/vsts/cli/work/_help.py",
-    "admin": "src/command_modules/vsts-cli-admin/vsts/cli/admin/_help.py",
-    "package": "src/command_modules/vsts-cli-package/vsts/cli/package/_help.py"
+  "build": "src/command_modules/vsts-cli-build/vsts/cli/build/_help.py",
+  "release": "src/command_modules/vsts-cli-release/vsts/cli/release/_help.py",
```
Comment (ishitam8) [inline scripts/docgen/doc_source_map.json]: Currently, folder name is **vrelease**. Hence this won't work.

Comment (nishubansal) [reply scripts/docgen/doc_source_map.json]: I am fixing the folder names. Its working in my local.

File: src/command_modules/vsts-cli-release/vsts/cli/release/arguments.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+def load_release_arguments(cli_command_loader):
+    with ArgumentsContext(cli_command_loader, 'release') as ac:
+        ac.argument('open_browser', options_list='--open')
+        ac.argument('project', options_list=('--project', '-p'))
+        ac.argument('team_instance', options_list=('--instance', '-i'))
+        ac.argument('detect', **enum_choice_list(_ON_OFF_SWITCH_VALUES))
+
+    with ArgumentsContext(cli_command_loader, 'release list') as ac:
+        ac.argument('definition_id', nargs='*', type=int)
```
Comment (atbagga) [inline src/command_modules/vsts-cli-release/vsts/cli/release/arguments.py]: Preferably add short names for definition-id, release-id, definition-name, release-name making it easier to type the commands. 

General rule is id and name of the main resource can be given by --name --id. e.g. release show command can have release-id as --id and release-name as --name
release definition show command can have definition-id as --id definition-name as --name.

File: src/common_modules/vsts-cli-release-common/vsts/cli/release/common/release_definition.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    """Opens the release definition in the default browser.
+    :param :class:`<ReleaseDefinitionReference> <release.v4_0.models.ReleaseDefinitionReference>` definition:
+    """
+    url = _get_release_definition_web_url(definition)
+    if url is not None and url:
+        logger.debug('Opening web page: %s', url)
+        open_new(url=url)
+
+
+def get_definition_id_from_name(name, client, project):
```
Comment (atbagga) [inline src/common_modules/vsts-cli-release-common/vsts/cli/release/common/release_definition.py]: nit: naming - Function for use in this file only. Use _get_definition_id_from_name. (_function_name)

Comment (nishubansal) [reply src/common_modules/vsts-cli-release-common/vsts/cli/release/common/release_definition.py]: This function is used in release.py file

**Comment by atbagga** [comment]
@nishubansal There was a [CI failure](https://mseng.visualstudio.com/vsts-cli/_build/results?buildId=7732866&view=logs) with this checkin. Please take a look.

**Pull request merged**